javalib/src/main/scala/java/lang/Math.scala

```diff
@@ -348,95 +348,136 @@ object Math {
-    multiplyExact(a, b.toLong)
+  def multiplyExact(a: scala.Long, b: scala.Int): scala.Long = {
+    /* Like multiplyExact(Long, Long), but note that b.toLong cannot be
+     * Long.MinValue, so we avoid some checks.
+     */
+    val bLong = b.toLong
+    val res = a * bLong
+    if (a != 0 && res / a != bLong)
+      longOverflow()
+    res
+  }
 
+  @inline
   def multiplyExact(a: scala.Long, b: scala.Long): scala.Long = {
-    val overflow = {
-      if (b > 0)
-        a > Long.MAX_VALUE / b || a < Long.MIN_VALUE / b
-      else if (b < -1)
-        a > Long.MIN_VALUE / b || a < Long.MAX_VALUE / b
-      else if (b == -1)
-        a == Long.MIN_VALUE
-      else
-        false
-    }
-    if (!overflow) a * b
-    else throw new ArithmeticException("Long overflow")
+    /* Hacker's Delight, Section 2-13
+     * We swap the role of a and b to match multiplyExact(Long, Int).
+     */
+    val res = a * b
+    if ((a < 0 && b == scala.Long.MinValue) || (a != 0 && res / a != b))
+      longOverflow()
+    res
   }
 
-  def incrementExact(a: scala.Int): scala.Int =
-    if (a != Integer.MAX_VALUE) a + 1
-    else throw new ArithmeticException("Integer overflow")
+  @inline
+  def incrementExact(a: scala.Int): scala.Int = {
+    if (a == Int.MaxValue)
+      intOverflow()
+    a + 1
+  }
 
-  def incrementExact(a: scala.Long): scala.Long =
-    if (a != Long.MAX_VALUE) a + 1
-    else throw new ArithmeticException("Long overflow")
+  @inline
+  def incrementExact(a: scala.Long): scala.Long = {
+    if (a == scala.Long.MaxValue)
+      longOverflow()
+    a + 1L
+  }
 
-  def decrementExact(a: scala.Int): scala.Int =
-    if (a != Integer.MIN_VALUE) a - 1
-    else throw new ArithmeticException("Integer overflow")
+  @inline
+  def decrementExact(a: scala.Int): scala.Int = {
+    if (a == Int.MinValue)
+      intOverflow()
+    a - 1
+  }
 
-  def decrementExact(a: scala.Long): scala.Long =
-    if (a != Long.MIN_VALUE) a - 1
-    else throw new ArithmeticException("Long overflow")
+  @inline
+  def decrementExact(a: scala.Long): scala.Long = {
+    if (a == scala.Long.MinValue)
+      longOverflow()
+    a - 1L
+  }
 
-  def negateExact(a: scala.Int): scala.Int =
-    if (a != Integer.MIN_VALUE) -a
-    else throw new ArithmeticException("Integer overflow")
+  @inline
+  def negateExact(a: scala.Int): scala.Int = {
+    if (a == Int.MinValue)
+      intOverflow()
+    -a
+  }
 
-  def negateExact(a: scala.Long): scala.Long =
-    if (a != Long.MIN_VALUE) -a
-    else throw new ArithmeticException("Long overflow")
+  @inline
+  def negateExact(a: scala.Long): scala.Long = {
+    if (a == scala.Long.MinValue)
+      longOverflow()
+    -a
+  }
 
-  def toIntExact(a: scala.Long): scala.Int =
-    if (a >= Integer.MIN_VALUE && a <= Integer.MAX_VALUE) a.toInt
-    else throw new ArithmeticException("Integer overflow")
+  @inline
+  def toIntExact(a: scala.Long): scala.Int = {
+    val res = a.toInt
+    if (res.toLong != a)
```

Comment:
Add a comment why this is better than checking against min / max value?

I realize the previous code had branches, but it seems they could easily be avoided (or do we not have a non short circuiting boolean and?).

Reply:
I added a comment. The long comparisons require 2 int comparisons each. With the new code, we only need 1 int comparison.